wit/types.wit

```diff
@@ -16,29 +25,56 @@ interface messaging-types {
-        format: format-spec,
-        metadata: option<list<tuple<string, string>>>
+    /// A message with a binary payload and additional information
+    resource message {
+        constructor(topic: channel, data: list<u8>, content-type: option<string>, metadata: option<list<tuple<string, string>>>);
+        /// The topic/subject/channel this message was received or should be sent on
+        topic: func() -> channel;
+        /// Set the topic/subject/channel this message should be sent on
+        set-topic: func(topic: channel);
+        /// An optional content-type describing the format of the data in the message. This is 
+        /// sometimes described as the "format" type
+        content-type: func() -> option<string>;
+        /// Set the content-type describing the format of the data in the message. This is
+        /// sometimes described as the "format" type
+        set-content-type: func(content-type: string);
+        /// An opaque blob of data
+        data: func() -> list<u8>;
+        /// Set the opaque blob of data for this message, discarding the old value
+        set-data: func(data: list<u8>);
+        /// Optional metadata (also called headers or attributes in some systems) attached to the
+        /// message
+        metadata: func() -> option<list<tuple<string, string>>>;
+        /// Add a new key-value pair to the metadata, overwriting any existing value for the same key
+        add-metadata: func(key: string, value: string);
+
+        /// Completes/acks the message
+        /// 
+        /// A message can exist under several statuses:
+        /// (1) available: the message is ready to be read,
+        /// (2) acquired: the message has been sent to a consumer (but still exists in the queue),
+        /// (3) accepted (result of complete): the message has been received and ACK-ed by a consumer and can be safely removed from the queue,
+        /// (4) rejected (result of abandon): the message has been received and NACK-ed by a consumer, at which point it can be:
+        ///         - deleted,
+        ///         - sent to a dead-letter queue, or
+        ///         - kept in the queue for further processing.
+        complete: func() -> result<_, error>;
+
+        /// Abandon/nacks the message
+        /// 
+        /// A message can exist under several statuses:
+        /// (1) available: the message is ready to be read,
+        /// (2) acquired: the message has been sent to a consumer (but still exists in the queue),
+        /// (3) accepted (result of complete): the message has been received and ACK-ed by a consumer and can be safely removed from the queue,
+        /// (4) rejected (result of abandon): the message has been received and NACK-ed by a consumer, at which point it can be:
+        ///         - deleted,
+        ///         - sent to a dead-letter queue, or
+        ///         - kept in the queue for further processing.
+        abandon: func() -> result<_, error>;
```

Comment:
```suggestion
        abandon: func(reason) -> result<_, error>;
```

I don't know if it would be worth adding some reason for abandonment. I'm just generalizing my thoughts here to see what sticks.
These would be friendly messages in the `Competing Consumer` model. 

Maybe it is not worth having it as a core functionality; people could add it as an add-on.

Action | Description
-- | --
Unknown | The client does not know what action to take. Let the server decide.
Park | Park the message and do not resend. Put it on poison queue.
Retry | Explicitly retry the message.
Skip | Skip this message do not resend and do not put in poison queue.

Reply:
AMQP structures related to this are here: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-rejected.

Reply:
What do we thing about having this just be a string rather than a specific state? I do like having additional context for sure, but am trying to keep this as simple as possible

Reply:
@thomastaylor312 yeah, that is why I left the type out.